river/ensemble/adaptive_random_forest.py

```diff
@@ -0,0 +1,824 @@
+            self.drift_detector = self.background_learner.drift_detector
+            self.metric = copy.deepcopy(self.background_learner.base_metric)
+            self.created_on = self.background_learner.created_on
+            self.background_learner = None
+        else:
+            # Reset model
+            self.model = copy.deepcopy(self.base_model)
+            self.metric = copy.deepcopy(self.base_metric)
+            self.created_on = n_samples_seen
+            self.drift_detector = copy.deepcopy(self.base_drift_detector)
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, *, sample_weight: int,   # noqa
+                  n_samples_seen: int):
+
+        self.model.learn_one(x, y, sample_weight=sample_weight)
+
+        if self.background_learner:
+            # Train the background learner
+            self.background_learner.model.learn_one(x=x, y=y, sample_weight=sample_weight)
+
+        if self._use_drift_detector and not self.is_background_learner:
+            drift_detector_input = self._drift_detector_input(
+                y_true=y, y_pred=self.model.predict_one(x)
+            )
+
+            # Check for warning only if use_background_learner is set
+            if self._use_background_learner:
+                self.warning_detector.update(drift_detector_input)
+                # Check if there was a (warning) change
+                if self.warning_detector.change_detected:
+                    self.last_warning_on = n_samples_seen
+                    self.n_warnings_detected += 1
+                    # Create a new background learner object
+                    self.background_learner = self.__class__(
+                        index_original=self.index_original,
+                        base_model=self.model.new_instance(),
+                        created_on=n_samples_seen,
+                        base_drift_detector=self.base_drift_detector,
+                        base_warning_detector=self.base_warning_detector,
+                        is_background_learner=True,
+                        base_metric=copy.deepcopy(self.base_metric)
+                    )
+                    # Reset the warning detector for the current object
+                    self.warning_detector = copy.deepcopy(self.base_warning_detector)
+
+            # Update the drift detector
+            self.drift_detector.update(drift_detector_input)
+
+            # Check if there was a change
+            if self.drift_detector.change_detected:
+                self.last_drift_on = n_samples_seen
+                self.n_drifts_detected += 1
+                self.reset(n_samples_seen)
+
+    @abc.abstractmethod
+    def _drift_detector_input(self,
+                              y_true: typing.Union[base.typing.ClfTarget, base.typing.RegTarget],
+                              y_pred: typing.Union[base.typing.ClfTarget, base.typing.RegTarget]):
+        raise NotImplementedError
```

Comment:
This is looking good!

Reply:
In future releases, this could be used to modify the classifier to track drift based on different metrics :-)